filters/tags_filter.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package filters
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+)
+
+// SortedTagIterator iterates over a set of tag names and values
+// sorted by tag names in ascending order
+type SortedTagIterator interface {
+	// Next returns true if there are more tag names and values
+	Next() bool
+
+	// Current returns the current tag name and value
+	Current() (string, string)
+
+	// Err returns any errors encountered
+	Err() error
+
+	// Close closes the iterator
+	Close()
+}
+
+// NewSortedTagIteratorFn creates a tag iterator given an id
+type NewSortedTagIteratorFn func(id string) SortedTagIterator
+
+// tagFilter is a filter associated with a given tag
+type tagFilter struct {
+	name        string
+	valueFilter Filter
+}
+
+func (f *tagFilter) String() string {
+	return fmt.Sprintf("%s:%s", f.name, f.valueFilter.String())
+}
+
+type tagFiltersByNameAsc []tagFilter
+
+func (tn tagFiltersByNameAsc) Len() int           { return len(tn) }
+func (tn tagFiltersByNameAsc) Swap(i, j int)      { tn[i], tn[j] = tn[j], tn[i] }
+func (tn tagFiltersByNameAsc) Less(i, j int) bool { return tn[i].name < tn[j].name }
+
+// tagsFilter contains a list of tag filters.
+type tagsFilter struct {
+	filters []tagFilter
+	iterFn  NewSortedTagIteratorFn
+	op      LogicalOp
```

Comment:
Do we foresee a usecase where "AND" is insufficient for the tags filter? M3QL only supports "AND" semantics when specifying tag name/value pairs essentially, so I'm not sure if we really need a logical op here. If the users need "OR" semantics, they could just create multiple rules, one for each tag name/value pair. If we keep the semantics to "AND" only, it reduces complexity in the filtering logic and the rule schema.